client/layout/masterbar/logged-in.jsx

```diff
@@ -145,30 +143,33 @@ const MasterbarLoggedIn = React.createClass( {
 // TODO: make this pure when sites can be retrieved from the Redux state
 export default connect( ( state, { sites } ) => {
 	let siteId = getSelectedSiteId( state );
-	let siteSlug = getSiteSlug( state, siteId );
 
-	// If siteId has not been set in redux, fall back to currentUser.primarySiteSlug
 	if ( ! siteId ) {
-		const currentUser = getCurrentUser( state );
-		siteSlug = get( currentUser, 'primarySiteSlug' );
-
-		if ( ! sites.getSite( siteSlug ) ) {
-			// The user's `primarySiteSlug` property might be stale if a site
-			// was just deleted, so we need to make sure the site is still in
-			// `sites-list`.
-			siteSlug = null;
-		}
+		// Falls back to using the user's primary site if no site has been selected in the user interface yet
+		siteId = get( getCurrentUser( state ), 'primary_blog' );
+	}
+
+	let siteSlug = getSiteSlug( state, siteId );
+
+	if ( siteSlug ) {
+		return {
+			siteSlug,
+			isDomainOnlySite: isDomainOnlySite( state, siteId )
+		};
+	} else {
+		let isDomainOnlySite = false;
 
-		// Now we can look up the site ID from its slug
-		const site = getSiteBySlug( state, siteSlug );
+		// Retrieves the site from the Sites store when the global state tree doesn't contain the list of sites yet
+		const site = sites.getSite( siteId );
 
 		if ( site ) {
-			siteId = site.ID;
+			siteSlug = site.slug;
+			isDomainOnlySite = get( site, 'options.is_domain_only', false );
```

Comment:
We use it above if we have retrieved the `siteSlug` from the global state tree successfully. However as mentioned in the comment in the code above the list of sites is not always available there: e.g. I've noticed that on page load. So the idea is to falls back to using the `Sites` store in that case.